ac-34 product exports

ExtraGravity opened the pull request.

ExtraGravity committed:
ac-34 product exports api docs

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *                - productTransactionUuid
+ *                - productType
+ *                - datetime (UTC)
+ *                - toPersonUuid
```
use apiparam tag pls :)

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *                - costPerUnit
+ *                - currency
+ *              Parameters have no effect on this request.
+ *              Associated attributes can be checked via /products/
```
example of associate attribute for product exports? also i think its a different end point for product exports?

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *              Associated attributes can be checked via /products/
+ * @apiName GetProductExports
+ *
+ * @apiError (401) Unauthorized - Must be admin
```
403

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+    {
+      "productTransactionUuid": "9f11efb5-d5b4-4853-aa88-ee10c2940c9f",
+      "datetime": "2018-01-23 04:05:06",
+      "toPersonUuid": "34e0574e-beea-498d-8bec-7f26d84ba761",
```
beea!

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+      "productUnits": "litres",
+      "costPerUnit": 22.56,
+      "currency": "UGX",
+      "density": "5"
```
maybe provide density units?

beaesguerra replied:
do we need density units?

ExtraGravity replied:
No, this is an extra attribute, so anything goes.

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
 }, authorized(UserType.ADMIN));
 
+/**
+ * @api {get} /productexports
```
should be post

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *                - datetime (UTC)
+ *                - productType
+ *                - toPersonUuid
+ *                - fromPersonUuid
```
api param please!

beaesguerra replied:
api param :)

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *                - costPerUnit
+ *                - currency
+ *              Along with any associated attributes.
+ *              Associated attributes can be checked via /products.
```
example?

beaesguerra replied:
wrong endpoint?

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *              The product UUID is returned.
+ * @apiName CreateProductExport
+ *
+ * @apiError (400) The following fields are missing, ...
```
specify it's a badrequest?

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *
+ * @apiError (400) The following fields are missing, ...
+ * @apiError (400) The product type does not exist
+ * @apiError (401) Unauthorized - Must be admin
```
forbidden

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *              Required fields are:
+ *                - productTransactionUuid
+ *                - productType
+ *                - datetime (UTC)
```
apiparam please!

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ *                - costPerUnit
+ *                - currency
+ *              Along with any associated attributes.
+ *              Associated attributes can be checked via /products.
```
wrong endpoint?

beaesguerra replied:
how are we getting associated attributes anyways for product exports?

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
+ * @apiName UpdateProductExport
+ *
+ * @apiError (400) The folllowing attributes do not exist
+ * @apiError (401) Unauthorized - Must be admin
```
403

beaesguerra commented on src/routers/products.ts:
```diff
@@ -0,0 +1,31 @@
+ *              Parameters have no effect on this request.
+ * @apiName GetPeople
+ *
+ * @apiError (401) Unauthorized - Must be admin
```
403

beaesguerra commented on src/routers/products.ts:
```diff
@@ -0,0 +1,31 @@
+  }
+ */
+router.get('/', async (req, res) => {
+  res.status(StatusCode.OK).send('Successfully retrieved all product exports') ;
```
should be products type not exports

ExtraGravity commented:
We don't need this yet so I'll fix it when we do.

ExtraGravity committed:
Merge branch 'master' into enoch/ac-34-product-exports-api-doc

ExtraGravity committed:
ac-34 Update product exports API docs

Docs should be up to standard of other docs.
People API docs have been fixed slightly to remove an error
that isn't being used.

ExtraGravity committed:
ac-34 fix linting errors

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -6,18 +6,107 @@ import { StatusCode } from '@/models/statusCodes';
+      "amountOfProduct": 5.2,
+      "productUnits": "kilograms",
+      "costPerUnit": 2.2,
+      "currency": "USD"
```
do you think we should be doing export products by type?

ExtraGravity replied:
Nah, not for now.

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -6,18 +6,107 @@ import { StatusCode } from '@/models/statusCodes';
 }, authorized(UserType.ADMIN));
 
+/**
+ * @api {post} /productexports
```
hehe i still think it would be nicer if it was /exports/products or products/exports or something so that its not multiple words. but that's just me!

ExtraGravity replied:
It does look nicer, but I don't think it makes as much sense. Since product exports is it's own entity.

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -6,18 +6,107 @@ import { StatusCode } from '@/models/statusCodes';
+ * @apiGroup ProductExports
+ * @apiVersion 0.0.1
+ * @apiDescription Updates a product export entry.
+ *              Possible fields are:
```
api params?

ExtraGravity replied:
Actually all the fields were wrong, must've been high.

beaesguerra reviewed (commented):
Re-reviewing. Some of my older comments havent been addressed still I think

ExtraGravity committed:
ac-34 update product exports api docs

ExtraGravity committed:
product exports get

ExtraGravity committed:
ac-34 post new product export and add last modified field to product exports

ExtraGravity committed:
ac-34 product transaction fixes and product export update

ExtraGravity committed:
Merge branch 'master' into enoch/ac-34-product-exports-api-doc

ExtraGravity commented:
Implementation is included in PR. But no testing with auth, couldn't get it to work even on other APIs.

ExtraGravity committed:
ac-34 fix linter errors

beaesguerra commented on dbscripts/db-init.sql:
```diff
@@ -208,8 +208,10 @@ CREATE TABLE ProductPayments (
 -- ProductExports
 
 CREATE TABLE ProductExports (
+	productExportUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
```
do we still need the uuid_generate_v4 if the client is sending us the uuid? 

don't have to do it as part of this change but i think since this has recently changed, we should be making sure in all of the requests that they provide the uuid (I don't think I have been checking for this in the past)

ExtraGravity replied:
Yea we shouldn't. I copy pasted most of this code from the ProductTransactions page which I hacked together as fast as I can.

beaesguerra commented on dbscripts/db-init.sql:
```diff
@@ -208,8 +208,10 @@ CREATE TABLE ProductPayments (
+	productExportUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
 	productTypeId SERIAL REFERENCES ProductTypes(productTypeId) NOT NULL,
 	amountOfProduct REAL NOT NULL,
 	transportId VARCHAR(255),
```
should this be not null?

beaesguerra commented on src/database/ProductExports/index.ts:
```diff
@@ -0,0 +1,83 @@
+/**
+ * Represent a product export after retrieval from the database
+ */
+export interface ProdExportDb {
```
i dont think this should have "db" after because productname isn't in the product export table?

ExtraGravity replied:
I think it's more like, what will the product exports database interface is gonna give. Not exactly what's in the database. Because this can't just return an ID, it's gotta give a product name somehow.

beaesguerra replied:
pls fix

ExtraGravity replied:
I don't remember what there was to fix.

beaesguerra commented on src/database/ProductExports/index.ts:
```diff
@@ -0,0 +1,83 @@
+}
+
+const builders = {
+  /** Get all product transactions of a certain type */
```
update comment

beaesguerra commented on src/database/ProductExports/index.ts:
```diff
@@ -0,0 +1,83 @@
+const builders = {
+  /** Get all product transactions of a certain type */
+  getSingleProdExport(productexportuuid: string) {
+    return prodExportTable().select('*')
```
does this need to return the product name? or is the id enough?

beaesguerra replied:
we should probably be consistent since the get all product exports api returns the productname too

ExtraGravity replied:
Good catch

beaesguerra commented on src/database/ProductExports/index.ts:
```diff
@@ -0,0 +1,83 @@
+      tableNames.PRODUCT_EXPORTS + '.producttypeid');
+  },
+
+  /** Inserts a new produt export row in the database */
```
typo in product

beaesguerra commented on dbscripts/db-init.sql:
```diff
@@ -208,8 +208,10 @@ CREATE TABLE ProductPayments (
 -- ProductExports
 
 CREATE TABLE ProductExports (
```
don't we need to track who exported the product? ie. which trader is adding it?

ExtraGravity replied:
This wasn't what was originally agreed to. Is this something we definitely want?

beaesguerra replied:
recorderUuid

beaesguerra commented on src/database/ProductExports/index.ts:
```diff
@@ -0,0 +1,83 @@
+  return exports;
+}
+
+/** creates a new product export in the database, returns the new UUID */
```
can we use proper js documentation? it usually auto generates it so that it requires you to specify what the params represent

beaesguerra replied:
and if there's a return value, it also lets you describe what its supposed to return

ExtraGravity replied:
This is alot of documentation to update haha.

beaesguerra commented on src/routers/index.ts:
```diff
@@ -13,5 +13,6 @@ router.use('/people', people);
 router.use('/products', products);
 router.use('/transactions', transactions);
 router.use('/productExports', productExports);
+router.use('/products', products);
```
this is already in line 13

beaesguerra commented on src/database/ProductExports/index.ts:
```diff
@@ -0,0 +1,83 @@
+
+/** creates a new product export in the database, returns the new UUID */
+export async function insertProdExport(req: ProdExportDbInsertReq): Promise<string> {
+  const newUuid = await execute<any>(builders.insertProdExport(req));
```
doesn't the server need to generate the last modified date?

beaesguerra replied:
ignore this, i saw it later on in the service file

beaesguerra commented on src/database/ProductExports/index.ts:
```diff
@@ -0,0 +1,83 @@
+
+/** Update a single column for a single product export */
+export async function updateProdExportField(uuid: string, field: string, value: string|number) {
+  return await execute<any>(builders.updateProdExportField(uuid, field, value));
```
doesn't the server need to generate the last modified date?

ExtraGravity replied:
Yup, missed it.

ExtraGravity replied:
Oh it's done in the service actually.

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -1,24 +1,171 @@
+  uuid: string;
+
+  transportId?: string;
+  datetime?: string;
```
we probably shouldn't let them update the datetime for when it was first created?

ExtraGravity replied:
I wanted to give the clients more freedom. There's alot of things that I think shouldn't be changeable but seems like people wanted anything to be changed, including money transactions.

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -1,24 +1,171 @@
+/**
+ * Represents a product export creation request in the API
+ */
+export interface ProdExportReq {
```
can you make this ProdExportCreationReq so it's clear and consistent with the other interfaces too?

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -1,24 +1,171 @@
-}, authorized(UserType.ADMIN));
+  const updateReq: ProdExportUpdateReq = {
+    uuid: req.params.uuid,
+
```
extra space

ExtraGravity replied:
I realized this was intentional to separate mandatory and non mandatory fields.

beaesguerra replied:
you could add a comment instead?

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -1,24 +1,171 @@
+    uuid: req.params.uuid,
+
+    transportId: req.body.transportId,
+    datetime: req.body.datetime,
```
they shouldn't be able to change datetime? but maybe the client should be generating the last modified timestamp?

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -1,24 +1,171 @@
+    await ProdExportsService.updateProdExportInDb(updateReq);
+  } catch (e) {
+    res.status(StatusCode.INTERNAL_SERVER_ERROR).send();
+  } finally {
```
should this be in a finally?  won't this still execute after an error is caught? how come this block of code can't be put in the try?

ExtraGravity replied:
Because the trader wants the actual correct info on error.

beaesguerra replied:
pls fix

beaesguerra replied:
pls fix

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -0,0 +1,134 @@
+ * Connector for the API and database to get a single product export
+ */
+export async function getProdExportFromDb(uuid: string): Promise<api.ProdExport> {
+  logger.info('getting single transaction', uuid);
```
did we want to keep this logging message?

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -0,0 +1,134 @@
+  const dbProdExports: db.ProdExportDb[] = await db.getProdExports();
+  const results: api.ProdExport[] = [];
+
+  logger.info('results', results);
```
did we want to log this? it'll always log an empty array

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -0,0 +1,134 @@
+      uuid: item.productexportuuid,
+      transportId: item.transportid,
+      datetime: item.datetime,
+      productType: item.productname,
```
if we're already returning the productname, should we also return the product units? it should be in the item already anyways since it was joined with the products table

ExtraGravity replied:
Sure

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -0,0 +1,134 @@
+  try {
+    productTypeId = await productTransactionsDb.getProductId(apiReq.productType);
+  } catch (e) {
+    throw new Error('Product type ' + apiReq.productType + ' not supported');
```
don't we also need to throw this error if producttypeid is still -1?

ExtraGravity replied:
I don't understand, the only time we need to throw this error is if we hit an error when we're getting the producttypeid.

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -0,0 +1,134 @@
+    invalidFields.push('datetime');
+  }
+  if (!apiReq.productType) {
+    invalidFields.push('productType');
```
would it be an invalid field if the product type isn't valid? should we be treating that case like how we are treating these cases?

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -0,0 +1,134 @@
+  if (!apiReq.amountOfProduct) {
+    invalidFields.push('amountOfProduct');
+  }
+
```
if we know all the invalid fields by this point, shouldnt we throw an error now instead of trying to insert a request?

ExtraGravity replied:
The error thrwoing is being done before the actual insert request is done. But I created the insert request struct before throwing the error, I'll switch them around.

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -0,0 +1,134 @@
+ */
+export async function updateProdExportInDb(apiReq: api.ProdExportUpdateReq) {
+  try {
+    if (apiReq.transportId) {
```
there's no way of doing an update of multiple fields in one query?

ExtraGravity replied:
Lazy :) 
I copied the code I hacked together. 

Let's optimize it later, we'll create a new issue for it and that way, muz can get some easy exposure to the backend that he was looking for.

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -0,0 +1,134 @@
+    if (apiReq.transportId) {
+      await db.updateProdExportField(apiReq.uuid, 'transportid', apiReq.transportId);
+    }
+    if (apiReq.datetime) {
```
we probabbly shouldn't allow this! :o

ExtraGravity committed:
ac-34 respond to bea's PR requests, add recorder uuid and not be able to edit datetime

ExtraGravity committed:
ac-34 respond to bea's PR requests

ExtraGravity committed:
ac-34 fix linting errors

beaesguerra commented on src/middleware/index.ts:
```diff
@@ -2,16 +2,14 @@ import * as compression from 'compression';
   this.use(createNetworkLogger());
-  this.use(jwtMiddleware());
-  this.use(unauthorizedMiddleware());
+  // this.use(jwtMiddleware());
```
put these back

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -33,9 +36,7 @@ export interface ProdExportReq {
   uuid: string;
-
   transportId?: string;
-  datetime?: string;
```
i thought we will let them change datetime? and shouldnt we also let them change the recorderuuid too?

ExtraGravity replied:
I forgot :P

beaesguerra commented on src/routers/productExports.ts:
```diff
@@ -131,8 +138,6 @@ router.post('/', async (req, res) => {
  * @apiParam {String} uuid Specify the UUID of the product export to update.
  * @apiParam [String] transportId Identifier for mode of transport, max 255 characters.
- * @apiParam [String] datetime The time the export occured.
- * @apiParam [String] productType The type of product.
```
why was this removed?

ExtraGravity replied:
They can't change the producttype

beaesguerra commented on src/services/ProductExports.ts:
```diff
@@ -119,9 +124,6 @@ export async function updateProdExportInDb(apiReq: api.ProdExportUpdateReq) {
     if (apiReq.transportId) {
       await db.updateProdExportField(apiReq.uuid, 'transportid', apiReq.transportId);
     }
-    if (apiReq.datetime) {
```
i thought we were letting them change datetime? and then should we let them change recorder uuid?

ExtraGravity committed:
ac-34 add back middleware, and editable datetime and recorderuuid
